File: src/lib/recommendation.ts
import { prisma } from "./db";

/**
 * Smart Recommendation Engine
 * Uses collaborative filtering based on reading history, tags, ratings, and genres.
 */

interface ScoredComic {
  id: string;
  title: string;
  score: number;
  reasons: string[];
  coverUrl: string;
  author: string;
  genre: string;
  tags: { name: string; color: string }[];
}

/**
 * Get personalized recommendations for a user
 */
export async function getRecommendations(options?: {
  limit?: number;
  excludeRead?: boolean;
}): Promise<ScoredComic[]> {
  const { limit = 10, excludeRead = false } = options || {};

  // Get all comics with their data
  const allComics = await prisma.comic.findMany({
    include: {
      tags: { include: { tag: true } },
      readingSessions: {
        select: { duration: true },
      },
    },
  });

  if (allComics.length === 0) return [];

  // Build user preference profile from reading behavior
  const profile = buildUserProfile(allComics);

  // Score each comic
  const scored: ScoredComic[] = [];

  for (const comic of allComics) {
    // Optionally exclude fully-read comics
    if (excludeRead && comic.lastReadPage > 0 && comic.pageCount > 0) {
      const progress = comic.lastReadPage / comic.pageCount;
      if (progress >= 0.9) continue;
    }

    const { score, reasons } = calculateScore(comic, profile);

    scored.push({
      id: comic.id,
      title: comic.title,
      score,
      reasons,
      coverUrl: `/api/comics/${comic.id}/thumbnail`,
      author: comic.author,
      genre: comic.genre,
      tags: comic.tags.map((ct) => ({ name: ct.tag.name, color: ct.tag.color })),
    });
  }

  // Sort by score descending
  scored.sort((a, b) => b.score - a.score);

  return scored.slice(0, limit);
}

interface UserProfile {
  // Tag preferences (tag -> weight)
  tagWeights: Map<string, number>;
  // Genre preferences (genre -> weight)
  genreWeights: Map<string, number>;
  // Author preferences (author -> weight)
  authorWeights: Map<string, number>;
  // Average rating given
  avgRating: number;
  // Series user is reading
  activeSeries: Set<string>;
  // Total reading time for normalization
  totalReadTime: number;
}

function buildUserProfile(
  comics: (typeof prisma.comic extends { findMany: (...args: unknown[]) => Promise<infer T> } ? (T extends (infer U)[] ? U : never) : never) extends infer C
    ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
      any[]
    : never
): UserProfile {
  const tagWeights = new Map<string, number>();
  const genreWeights = new Map<string, number>();
  const authorWeights = new Map<string, number>();
  const activeSeries = new Set<string>();
  let totalRating = 0;
  let ratedCount = 0;
  let totalReadTime = 0;

  for (const comic of comics) {
    // Weight based on engagement (read time, rating, favorite)
    const engagement = calculateEngagement(comic);
    if (engagement <= 0) continue;

    // Tag weights
    if (comic.tags) {
      for (const ct of comic.tags) {
        const tagName = ct.tag?.name || ct.tag;
        const current = tagWeights.get(tagName) || 0;
        tagWeights.set(tagName, current + engagement);
      }
    }

    // Genre weights
    if (comic.genre) {
      const genres = comic.genre.split(",").map((g: string) => g.trim()).filter(Boolean);
      for (const genre of genres) {
        const current = genreWeights.get(genre) || 0;
        genreWeights.set(genre, current + engagement);
      }
    }

    // Author weights
    if (comic.author) {
      const current = authorWeights.get(comic.author) || 0;
      authorWeights.set(comic.author, current + engagement);
    }

    // Rating stats
    if (comic.rating) {
      totalRating += comic.rating;
      ratedCount++;
    }

    // Active series
    if (comic.seriesName && comic.lastReadPage > 0) {
      activeSeries.add(comic.seriesName);
    }

    totalReadTime += comic.totalReadTime || 0;
  }

  return {
    tagWeights,
    genreWeights,
    authorWeights,
    avgRating: ratedCount > 0 ? totalRating / ratedCount : 3,
    activeSeries,
    totalReadTime,
  };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function calculateEngagement(comic: any): number {
  let score = 0;

  // Reading time contributes to engagement
  const readTime = comic.totalReadTime || 0;
  if (readTime > 0) score += Math.min(readTime / 600, 5); // Cap at 5 for 10 min

  // Reading progress
  if (comic.pageCount > 0 && comic.lastReadPage > 0) {
    const progress = comic.lastReadPage / comic.pageCount;
    score += progress * 3; // 0-3 points for progress
  }

  // Rating boost
  if (comic.rating) {
    score += (comic.rating - 2.5) * 2; // -5 to +5 based on rating
  }

  // Favorite boost
  if (comic.isFavorite) {
    score += 3;
  }

  // Recency boost
  if (comic.lastReadAt) {
    const daysSince = (Date.now() - new Date(comic.lastReadAt).getTime()) / (1000 * 60 * 60 * 24);
    if (daysSince < 7) score += 2;
    else if (daysSince < 30) score += 1;
  }

  return score;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function calculateScore(comic: any, profile: UserProfile): { score: number; reasons: string[] } {
  let score = 0;
  const reasons: string[] = [];

  // 1. Tag match (max 30 points)
  let tagScore = 0;
  if (comic.tags) {
    for (const ct of comic.tags) {
      const tagName = ct.tag?.name || ct.tag;
      const weight = profile.tagWeights.get(tagName) || 0;
      tagScore += weight;
    }
  }
  if (tagScore > 0) {
    const normalizedTag = Math.min(tagScore / 10, 30);
    score += normalizedTag;
    if (normalizedTag > 5) reasons.push("tag_match");
  }

  // 2. Genre match (max 25 points)
  if (comic.genre) {
    let genreScore = 0;
    const genres = comic.genre.split(",").map((g: string) => g.trim());
    for (const genre of genres) {
      genreScore += profile.genreWeights.get(genre) || 0;
    }
    if (genreScore > 0) {
      const normalizedGenre = Math.min(genreScore / 10, 25);
      score += normalizedGenre;
      if (normalizedGenre > 5) reasons.push("genre_match");
    }
  }

  // 3. Author match (max 20 points)
  if (comic.author) {
    const authorWeight = profile.authorWeights.get(comic.author) || 0;
    if (authorWeight > 0) {
      const normalizedAuthor = Math.min(authorWeight / 5, 20);
      score += normalizedAuthor;
      reasons.push("same_author");
    }
  }

  // 4. Series continuation bonus (max 15 points)
  if (comic.seriesName && profile.activeSeries.has(comic.seriesName)) {
    // Unread entries in a series you're reading
    if (comic.pageCount > 0 && comic.lastReadPage === 0) {
      score += 15;
      reasons.push("series_continuation");
    } else if (comic.lastReadPage > 0 && comic.lastReadPage < comic.pageCount * 0.9) {
      score += 10;
      reasons.push("series_in_progress");
    }
  }

  // 5. Rating prediction bonus (max 10 points)
  if (comic.rating && comic.rating >= profile.avgRating) {
    score += (comic.rating - profile.avgRating) * 3;
    reasons.push("highly_rated");
  }

  // 6. Unread bonus (encourage discovery)
  if (comic.lastReadPage === 0 && comic.pageCount > 0) {
    score += 5;
    reasons.push("unread");
  }

  // 7. Recency penalty (recently read = lower priority for "next read")
  if (comic.lastReadAt) {
    const daysSince = (Date.now() - new Date(comic.lastReadAt).getTime()) / (1000 * 60 * 60 * 24);
    if (daysSince < 1) score -= 10;
    else if (daysSince < 3) score -= 5;
  }

  return { score: Math.max(0, score), reasons };
}

/**
 * Get "similar comics" for a given comic
 */
export async function getSimilarComics(comicId: string, limit = 5): Promise<ScoredComic[]> {
  const target = await prisma.comic.findUnique({
    where: { id: comicId },
    include: { tags: { include: { tag: true } } },
  });

  if (!target) return [];

  const allComics = await prisma.comic.findMany({
    where: { id: { not: comicId } },
    include: {
      tags: { include: { tag: true } },
    },
  });

  const targetTags = new Set(target.tags.map((ct) => ct.tag.name));
  const targetGenres = new Set(
    (target.genre || "").split(",").map((g) => g.trim()).filter(Boolean)
  );

  const scored: ScoredComic[] = [];

  for (const comic of allComics) {
    let score = 0;
    const reasons: string[] = [];

    // Tag overlap (Jaccard similarity)
    const comicTags = new Set(comic.tags.map((ct) => ct.tag.name));
    const intersection = [...targetTags].filter((t) => comicTags.has(t)).length;
    const union = new Set([...targetTags, ...comicTags]).size;
    if (union > 0) {
      const tagSimilarity = intersection / union;
      score += tagSimilarity * 40;
      if (tagSimilarity > 0.3) reasons.push("similar_tags");
    }

    // Genre overlap
    const comicGenres = new Set(
      (comic.genre || "").split(",").map((g) => g.trim()).filter(Boolean)
    );
    const genreIntersection = [...targetGenres].filter((g) => comicGenres.has(g)).length;
    const genreUnion = new Set([...targetGenres, ...comicGenres]).size;
    if (genreUnion > 0) {
      const genreSimilarity = genreIntersection / genreUnion;
      score += genreSimilarity * 30;
      if (genreSimilarity > 0.3) reasons.push("similar_genre");
    }

    // Same author
    if (comic.author && comic.author === target.author) {
      score += 20;
      reasons.push("same_author");
    }

    // Same series
    if (comic.seriesName && comic.seriesName === target.seriesName) {
      score += 25;
      reasons.push("same_series");
    }

    // Same group
    if (comic.groupName && comic.groupName === target.groupName) {
      score += 10;
      reasons.push("same_group");
    }

    if (score > 0) {
      scored.push({
        id: comic.id,
        title: comic.title,
        score,
        reasons,
        coverUrl: `/api/comics/${comic.id}/thumbnail`,
        author: comic.author,
        genre: comic.genre,
        tags: comic.tags.map((ct) => ({ name: ct.tag.name, color: ct.tag.color })),
      });
    }
  }

  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, limit);
}

File: internal/service/recommendation.go
package service

import (
	"fmt"
	"math"
	"math/rand"
	"sort"
	"strings"
	"time"

	"github.com/nowen-reader/nowen-reader/internal/store"
)

// ScoredComic represents a recommended comic with scoring details.
type ScoredComic struct {
	ID       string               `json:"id"`
	Title    string               `json:"title"`
	Score    float64              `json:"score"`
	Reasons  []string             `json:"reasons"`
	AIReason string               `json:"aiReason,omitempty"` // AI 生成的自然语言推荐理由
	CoverURL string               `json:"coverUrl"`
	Author   string               `json:"author"`
	Genre    string               `json:"genre"`
	Filename string               `json:"filename"`
	Tags     []store.ComicTagInfo `json:"tags"`
}

// GetRecommendations returns personalized comic recommendations.
// seed > 0 时会在评分上添加随机扰动，使每次刷新结果不同。
func GetRecommendations(limit int, excludeRead bool, contentType string, seed int64) ([]ScoredComic, error) {
	allComics, err := store.GetAllComicsForRecommendation()
	if err != nil || len(allComics) == 0 {
		return []ScoredComic{}, nil
	}

	profile := buildUserProfile(allComics)

	// 根据 seed 创建随机源（seed==0 时不添加随机扰动）
	var rng *rand.Rand
	if seed > 0 {
		rng = rand.New(rand.NewSource(seed))
	}

	var scored []ScoredComic
	for _, comic := range allComics {
		// 按内容类型过滤
		if contentType == "novel" && !IsNovelFilename(comic.Filename) {
			continue
		}
		if contentType == "comic" && IsNovelFilename(comic.Filename) {
			continue
		}

		if excludeRead && comic.LastReadPage > 0 && comic.PageCount > 0 {
			progress := float64(comic.LastReadPage) / float64(comic.PageCount)
			if progress >= 0.9 {
				continue
			}
		}

		score, reasons := calculateRecommendationScore(comic, profile)

		// 添加随机扰动（±20%），使刷新结果有变化
		if rng != nil && score > 0 {
			jitter := 0.8 + rng.Float64()*0.4 // 0.8 ~ 1.2
			score *= jitter
		}

		scored = append(scored, ScoredComic{
			ID:       comic.ID,
			Title:    comic.Title,
			Score:    score,
			Reasons:  reasons,
			CoverURL: fmt.Sprintf("/api/comics/%s/thumbnail", comic.ID),
			Author:   comic.Author,
			Genre:    comic.Genre,
			Filename: comic.Filename,
			Tags:     comic.Tags,
		})
	}

	// 使用 sort.Slice 替代冒泡排序，O(n²) → O(n log n)
	sort.Slice(scored, func(i, j int) bool {
		return scored[i].Score > scored[j].Score
	})

	if limit > 0 && len(scored) > limit {
		scored = scored[:limit]
	}
	return scored, nil
}

// GetSimilarComics returns comics similar to a given comic.
func GetSimilarComics(comicID string, limit int) ([]ScoredComic, error) {
	target, err := store.GetComicByID(comicID)
	if err != nil || target == nil {
		return []ScoredComic{}, nil
	}

	allComics, err := store.GetAllComicsForRecommendation()
	if err != nil {
		return nil, err
	}

	targetTags := map[string]bool{}
	for _, t := range target.Tags {
		targetTags[t.Name] = true
	}
	targetGenres := map[string]bool{}
	for _, g := range strings.Split(target.Genre, ",") {
		g = strings.TrimSpace(g)
		if g != "" {
			targetGenres[g] = true
		}
	}
	targetCats := map[string]bool{}
	for _, c := range target.Categories {
		targetCats[c.Slug] = true
	}

	var scored []ScoredComic
	for _, comic := range allComics {
		if comic.ID == comicID {
			continue
		}

		var score float64
		var reasons []string

		// Tag overlap (Jaccard)
		comicTags := map[string]bool{}
		for _, t := range comic.Tags {
			comicTags[t.Name] = true
		}
		intersection := 0
		for t := range targetTags {
			if comicTags[t] {
				intersection++
			}
		}
		unionSize := len(targetTags)
		for t := range comicTags {
			if !targetTags[t] {
				unionSize++
			}
		}
		if unionSize > 0 {
			tagSim := float64(intersection) / float64(unionSize)
			score += tagSim * 40
			if tagSim > 0.3 {
				reasons = append(reasons, "similar_tags")
			}
		}

		// Genre overlap
		comicGenres := map[string]bool{}
		for _, g := range strings.Split(comic.Genre, ",") {
			g = strings.TrimSpace(g)
			if g != "" {
				comicGenres[g] = true
			}
		}
		genreIntersection := 0
		for g := range targetGenres {
			if comicGenres[g] {
				genreIntersection++
			}
		}
		genreUnion := len(targetGenres)
		for g := range comicGenres {
			if !targetGenres[g] {
				genreUnion++
			}
		}
		if genreUnion > 0 {
			genreSim := float64(genreIntersection) / float64(genreUnion)
			score += genreSim * 30
			if genreSim > 0.3 {
				reasons = append(reasons, "similar_genre")
			}
		}

		// Same author
		if comic.Author != "" && comic.Author == target.Author {
			score += 20
			reasons = append(reasons, "same_author")
		}

		// Same category
		for _, c := range comic.Categories {
			if targetCats[c.Slug] {
				score += 8
				reasons = append(reasons, "same_category")
				break
			}
		}

		if score > 0 {
			scored = append(scored, ScoredComic{
				ID:       comic.ID,
				Title:    comic.Title,
				Score:    score,
				Reasons:  reasons,
				CoverURL: fmt.Sprintf("/api/comics/%s/thumbnail", comic.ID),
				Author:   comic.Author,
				Genre:    comic.Genre,
				Filename: comic.Filename,
				Tags:     comic.Tags,
			})
		}
	}

	// 使用 sort.Slice 替代冒泡排序，O(n²) → O(n log n)
	sort.Slice(scored, func(i, j int) bool {
		return scored[i].Score > scored[j].Score
	})

	if limit > 0 && len(scored) > limit {
		scored = scored[:limit]
	}
	return scored, nil
}

// ============================================================
// Internal
// ============================================================

type userProfile struct {
	tagWeights    map[string]float64
	genreWeights  map[string]float64
	authorWeights map[string]float64
	avgRating     float64
}

func buildUserProfile(comics []store.RecommendationComic) userProfile {
	p := userProfile{
		tagWeights:    map[string]float64{},
		genreWeights:  map[string]float64{},
		authorWeights: map[string]float64{},
	}

	var totalRating float64
	var ratedCount int

	for _, c := range comics {
		engagement := calculateEngagement(c)
		if engagement <= 0 {
			continue
		}

		for _, t := range c.Tags {
			p.tagWeights[t.Name] += engagement
		}

		if c.Genre != "" {
			for _, g := range strings.Split(c.Genre, ",") {
				g = strings.TrimSpace(g)
				if g != "" {
					p.genreWeights[g] += engagement
				}
			}
		}

		if c.Author != "" {
			p.authorWeights[c.Author] += engagement
		}

		if c.Rating != nil {
			totalRating += float64(*c.Rating)
			ratedCount++
		}
	}

	if ratedCount > 0 {
		p.avgRating = totalRating / float64(ratedCount)
	} else {
		p.avgRating = 3
	}
	return p
}

func calculateEngagement(c store.RecommendationComic) float64 {
	var score float64

	readTime := c.TotalReadTime
	if readTime > 0 {
		score += math.Min(float64(readTime)/600, 5)
	}

	if c.PageCount > 0 && c.LastReadPage > 0 {
		progress := float64(c.LastReadPage) / float64(c.PageCount)
		score += progress * 3
	}

	if c.Rating != nil {
		score += (float64(*c.Rating) - 2.5) * 2
	}

	if c.IsFavorite {
		score += 3
	}

	if c.LastReadAt != nil {
		daysSince := time.Since(*c.LastReadAt).Hours() / 24
		if daysSince < 7 {
			score += 2
		} else if daysSince < 30 {
			score += 1
		}
	}

	return score
}

func calculateRecommendationScore(c store.RecommendationComic, profile userProfile) (float64, []string) {
	var score float64
	var reasons []string

	// Tag match
	var tagScore float64
	for _, t := range c.Tags {
		tagScore += profile.tagWeights[t.Name]
	}
	if tagScore > 0 {
		normalized := math.Min(tagScore/10, 30)
		score += normalized
		if normalized > 5 {
			reasons = append(reasons, "tag_match")
		}
	}

	// Genre match
	if c.Genre != "" {
		var genreScore float64
		for _, g := range strings.Split(c.Genre, ",") {
			g = strings.TrimSpace(g)
			genreScore += profile.genreWeights[g]
		}
		if genreScore > 0 {
			normalized := math.Min(genreScore/10, 25)
			score += normalized
			if normalized > 5 {
				reasons = append(reasons, "genre_match")
			}
		}
	}

	// Author match
	if c.Author != "" {
		if w := profile.authorWeights[c.Author]; w > 0 {
			normalized := math.Min(w/5, 20)
			score += normalized
			reasons = append(reasons, "same_author")
		}
	}

	// Rating prediction
	if c.Rating != nil && float64(*c.Rating) >= profile.avgRating {
		score += (float64(*c.Rating) - profile.avgRating) * 3
		reasons = append(reasons, "highly_rated")
	}

	// Unread bonus
	if c.LastReadPage == 0 && c.PageCount > 0 {
		score += 5
		reasons = append(reasons, "unread")
	}

	// Recency penalty
	if c.LastReadAt != nil {
		daysSince := time.Since(*c.LastReadAt).Hours() / 24
		if daysSince < 1 {
			score -= 10
		} else if daysSince < 3 {
			score -= 5
		}
	}

	if score < 0 {
		score = 0
	}
	if reasons == nil {
		reasons = []string{}
	}
	return score, reasons
}

// IsNovelFilename 判断文件名是否为小说格式
func IsNovelFilename(filename string) bool {
	lower := strings.ToLower(filename)
	return strings.HasSuffix(lower, ".txt") ||
		strings.HasSuffix(lower, ".epub") ||
		strings.HasSuffix(lower, ".mobi") ||
		strings.HasSuffix(lower, ".azw3")
}
